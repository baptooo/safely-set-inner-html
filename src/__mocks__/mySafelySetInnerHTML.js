import SafelySetInnerHTML from '../safelySetInnerHTML';

const instance = new SafelySetInnerHTML();

export default instance.transform;
