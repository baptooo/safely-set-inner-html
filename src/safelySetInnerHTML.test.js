import React from 'react';
import renderer from 'react-test-renderer';
import SafelySetInnerHTML from './safelySetInnerHTML';
import mySafelySetInnerHTML from './__mocks__/mySafelySetInnerHTML';

jest.spyOn(console, 'warn');

describe('SafelySetInnerHTML', () => {
  describe('initialization', () => {
    let config = { ALLOWED_TAGS: ['marquee'] };
    let instance = new SafelySetInnerHTML(config);

    it('should merge the config', () => {
      expect(instance.config).toEqual({
        ...SafelySetInnerHTML.defaultConfig,
        ...config
      });
    });

    it('should set tagId to 0', () => {
      expect(instance.tagId).toEqual(0);
    });

    it('should init cache', () => {
      expect(instance.cache).toEqual([]);
    });
  });

  describe('rendering process', () => {
    it('renders correctly', () => {
      const instance = new SafelySetInnerHTML();

      const dom = instance.transform('Hello <strong>World !</strong>');
      const tree = renderer.create(<p>{dom}</p>).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders only allowedTags', () => {
      const instance = new SafelySetInnerHTML({ ALLOWED_TAGS: [] });

      const dom = instance.transform('Hello <strong>World !</strong>');
      const tree = renderer.create(<p>{dom}</p>).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders correctly', () => {
      const instance = new SafelySetInnerHTML();

      const dom = instance.transform('About the <a href="http://example.com">Author !</a>');
      const tree = renderer.create(<cite>{dom}</cite>).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('attributes rules', () => {
    it('should format the attributes properly', () => {
      const instance = new SafelySetInnerHTML({
        ALLOWED_ATTRIBUTES: ['href']
      });

      const props = instance.formatAttributes([
        { key: 'foo', value: 'bar' },
        { key: 'href', value: 'http://example.com' }
      ]);

      expect(props).toEqual({
        href: 'http://example.com'
      })
    });

    it('should work with React special properties', () => {
      const instance = new SafelySetInnerHTML({
        ALLOWED_TAGS: ['article', 'h2', 'p'],
        ALLOWED_ATTRIBUTES: ['class']
      });

      const dom = instance.transform(`
      <article class="article">
        <h2 class="article__title">Article title</h2>      
        <p class="article__content">Lorem ipsum dolor sit amet</p>      
      </article>
    `);
      const tree = renderer.create(dom).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('XSS attack prevention', () => {
    /**
     * This case can appear irrelevant but for some cases such as emails
     * it could happen to you
     */
    it('should prevent basic XSS attacks', () => {
      const instance = new SafelySetInnerHTML({
        ALLOWED_TAGS: ['style', 'div'],
        ALLOWED_ATTRIBUTES: ['']
      });

      const dom = instance.transform(`
      <style type="text/css">
        [ontransitionend] {
          transition: all 0.2s;
        }
        [ontransitionend]:hover {
          color: black;
        }
      </style>
      <div ontransitionend="console.log('xss');">XSS Attack</div>
    `);
      const tree = renderer.create(<div>{dom}</div>).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should warn the user about potential XSS attack', () => {
      const instance = new SafelySetInnerHTML({
        ALLOWED_TAGS: ['div'],
        ALLOWED_ATTRIBUTES: ['ontransitionend']
      });

      instance.transform('<div ontransitionend="alert(true)">Warn me !</div>');

      expect(console.warn).toHaveBeenCalled();
    });
  });

  describe('cache', () => {
    let cacheInstance = new SafelySetInnerHTML();
    let cacheStr = 'Hello <strong>Cache !</strong>';
    let generateDomSpy;

    beforeEach(() => {
      generateDomSpy = jest.spyOn(cacheInstance, 'generateDom');
    });

    afterEach(() => {
      generateDomSpy.mockReset();
      generateDomSpy.mockRestore();
    });

    it('should render normally and store in cache', () => {
      cacheInstance.transform(cacheStr);
      const [cachedString] = cacheInstance.cache;

      expect(cacheInstance.cache).toHaveLength(1);
      expect(cachedString.str).toEqual(cacheStr);
      expect(cacheInstance.generateDom).toHaveBeenCalled();
    });

    it('should retrieve dom from cache and prevent dom regeneration', () => {
      cacheInstance.transform(cacheStr);
      const [cachedString] = cacheInstance.cache;

      expect(cacheInstance.cache).toHaveLength(1);
      expect(cachedString.str).toEqual(cacheStr);
      expect(cacheInstance.generateDom).not.toHaveBeenCalled();
    });
  });

  describe('configuration file', () => {
    it('should render normally', () => {
      const dom = mySafelySetInnerHTML('Configuration <strong>File !</strong>');
      const tree = renderer.create(<p>{dom}</p>).toJSON();

      expect(tree).toMatchSnapshot();
    })
  });

  describe('empty elements', () => {
    it('should render br', () => {
      const instance = new SafelySetInnerHTML({ ALLOWED_TAGS: ['br'] });
      const dom = instance.transform('Hello <br /> there');
      const tree = renderer.create(<p>{dom}</p>).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render input', () => {
      const instance = new SafelySetInnerHTML({ ALLOWED_TAGS: ['input'] });
      const dom = instance.transform('<input type="text" />');
      const tree = renderer.create(<p>{dom}</p>).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
