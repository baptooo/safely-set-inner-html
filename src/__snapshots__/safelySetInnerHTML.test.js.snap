// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SafelySetInnerHTML HTML entities should render Espace&nbsp;insécable properly 1`] = `
<p>
  Espace insécable
</p>
`;

exports[`SafelySetInnerHTML HTML entities should render L&#8217;arbre est vert properly 1`] = `
<p>
  L’arbre est vert
</p>
`;

exports[`SafelySetInnerHTML XSS attack prevention should prevent basic XSS attacks 1`] = `
<div>
  
      
  <style>
    
        [ontransitionend] {
          transition: all 0.2s;
        }
        [ontransitionend]:hover {
          color: black;
        }
      
  </style>
  
      
  <div>
    XSS Attack
  </div>
  
    
</div>
`;

exports[`SafelySetInnerHTML attributes rules [href] should prevent href with javascript call 1`] = `
<a>
  Foo
</a>
`;

exports[`SafelySetInnerHTML attributes rules [href] should prevent href with spaces and javascript call 1`] = `
<a>
  Foo
</a>
`;

exports[`SafelySetInnerHTML attributes rules should work with React special properties 1`] = `
Array [
  "
      ",
  <article
    class="article"
  >
    
        
    <h2
      class="article__title"
    >
      Article title
    </h2>
          
        
    <p
      class="article__content"
    >
      Lorem ipsum dolor sit amet
    </p>
          
      
  </article>,
  "
    ",
]
`;

exports[`SafelySetInnerHTML configuration file should render normally 1`] = `
<p>
  Configuration 
  <strong>
    File !
  </strong>
</p>
`;

exports[`SafelySetInnerHTML empty elements should render br 1`] = `
<p>
  Hello 
  <br />
   there
</p>
`;

exports[`SafelySetInnerHTML empty elements should render input 1`] = `
<p>
  <input />
</p>
`;

exports[`SafelySetInnerHTML rendering process renders correctly 1`] = `
<p>
  Hello 
  <strong>
    World !
  </strong>
</p>
`;

exports[`SafelySetInnerHTML rendering process renders correctly 2`] = `
<cite>
  About the 
  <a
    href="http://example.com"
  >
    Author !
  </a>
</cite>
`;

exports[`SafelySetInnerHTML rendering process renders only allowedTags 1`] = `
<p>
  Hello 
  World !
</p>
`;
